### PR TITLE
Add support for async and generator functions in rx.pipe

### PR DESCRIPTION
**Pull request by philippjfr**
We already support using asynchronous and generator functions as inputs to a reactive expression but until now we did not handle them when you use them as an input to an operation that is performed on the data. This meant that `.rx.pipe` and `.rx.map` would simply return the coroutine or generator as the value instead of evaluating it lazily.

Here we add support for these by scheduling the task on the event loop and then triggering an event when the function evaluates. Internally this is achieved by creating a `Trigger` which is added as one of the dependencies of the reactive expression such as that any consumer of the expression (whether Panel, the display hook or anything else) is notified when the async function or generator yields a new value.

In testing this I also found an oversight which meant that if you used a bound function as the input to `.rx.pipe` it would not extract its dependencies.

### CODE REVIEW

File: param/reactive.py [hunk 1]
```diff
@@ -256,8 +258,12 @@ def map(self, func, /, *args, **kwargs):
         kwargs: mapping, optional
           A dictionary of keywords to pass to `func`.
         """
-        def apply(vs, *args, **kwargs):
-            return [func(v, *args, **kwargs) for v in vs]
+        if inspect.iscoroutinefunction(func):
```
Comment (Hoxbro) [inline param/reactive.py:261]: Should we raise an exception if a generator is used for maps? I don't think it makes much sense.

File: tests/testreactive.py [hunk 2]
```diff
@@ -551,6 +576,24 @@ def gen():
...
+    rxgen = rxv.rx.pipe(gen)
+    rxgen.rx.watch()
+    assert rxgen.rx.value is param.Undefined
+    await asyncio.sleep(0.04)
+    assert rxgen.rx.value == 1
+    await asyncio.sleep(0.1)
+    assert rxgen.rx.value == 2
+    rxv.rx.value = 2
+    await asyncio.sleep(0.03)
+    assert rxgen.rx.value == 3
```
Comment (Hoxbro) [inline tests/testreactive.py:595]: I'm not entirely sure why the first one is 0.04, and the last one is 0.03. Is that just because it works?

**Review by Hoxbro — COMMENTED** [review]
Some minor comments, but LGTM.

File: tests/testreactive.py (hunk 2 above)
Comment (philippjfr) [reply tests/testreactive.py:595]: Solid question, really we need a `wait_until` utility like we have in Panel.

File: param/reactive.py (hunk 1 above)
Comment (philippjfr) [reply param/reactive.py:261]: Agreed, I almost went down the road of implementing it but it was just a little too crazy to be sensible.

File: tests/testreactive.py (hunk 2 above)
Comment (Hoxbro) [reply tests/testreactive.py:595]: Steal it

**Pull request merged**